src/2013-07-13-dmvnorm_arma.Rmd

```diff
@@ -34,47 +34,128 @@ dMvn <- function(X,mu,Sigma) {
+                      arma::mat const &sigma, 
+                      bool const logd = false) { 
+    using arma::uword;
+    uword const n = x.n_rows, 
+             xdim = x.n_cols;
+    arma::vec out(n);
+    arma::mat const rooti = arma::inv(trimatu(arma::chol(sigma)));
+    double const rootisum = arma::sum(log(rooti.diag())), 
+                constants = -(double)xdim/2.0 * log2pi, 
+              other_terms = rootisum + constants;
+    
+    arma::rowvec z;
+    for (uword i = 0; i < n; i++) {
+        z      = (x.row(i) - mean) * rooti;    
+        out(i) = other_terms - 0.5 * arma::dot(z, z);     
+    }  
+      
+    if (logd)
+      return out;
+    return exp(out);
+}
+
+/* C++ version of the dtrmv BLAS function */
+void inplace_tri_mat_mult(arma::rowvec &x, arma::mat const &trimat){
+  arma::uword const n = trimat.n_cols;
+  
+  for(unsigned j = n; j-- > 0;){
+    double tmp(0.);
+    for(unsigned i = 0; i <= j; ++i)
+      tmp += trimat.at(i, j) * x[i];
+    x[j] = tmp;
+  }
+}
+
+// [[Rcpp::export]]
+arma::vec dmvnrm_arma_fast(arma::mat const &x,  
+                           arma::rowvec const &mean,  
+                           arma::mat const &sigma, 
+                           bool const logd = false) { 
+    using arma::uword;
+    uword const n = x.n_rows, 
+             xdim = x.n_cols;
+    arma::vec out(n);
+    arma::mat const rooti = arma::inv(trimatu(arma::chol(sigma)));
+    double const rootisum = arma::sum(log(rooti.diag())), 
+                constants = -(double)xdim/2.0 * log2pi, 
+              other_terms = rootisum + constants;
+    
+    arma::rowvec z;
+    for (uword i = 0; i < n; i++) {
+        z = (x.row(i) - mean);
+        inplace_tri_mat_mult(z, rooti);
+        out(i) = other_terms - 0.5 * arma::dot(z, z);     
+    }  
+      
+    if (logd)
+      return out;
+    return exp(out);
 }
 ```
 
+The use of `trimatu` allows to exploit that of the Cholesky
+decomposition the covariance matrix is an upper triangular matrix in the 
+inversion. The `dmvnrm_arma_old` is an older version of the function used in
+a previous version of this article. The new version differs mainly by
+
+ 1. using `const &` for the input parameters. 
+ 2. declaring `z` outside the loop.  
+ 3. using `arma::dot` instead of `arma::sum`. 
+ 4. other minor things.
+ 
+This turns out to be quite important for the computation times. The 
+`dmvnrm_arma_fast` makes an inplace vector matrix product and exploits
+that the matrix is an upper triangular matrix. One can use the 
+[dtrmv](http://www.netlib.org/lapack/explore-html/d7/d15/group__double__blas__level2_ga596c2acd9f81df6608bd5ed97e193897.html#ga596c2acd9f81df6608bd5ed97e193897)
+BLAS function instead. It is not available through the Armadillo library 
+though.
+
 Additionally, we can make use of the OpenMP library to use multiple 
 cores. For the OpenMP implementation, we need to enable OpenMP support. 
 One way of doing so is by adding the required compiler and linker 
```

Comment:
It's been a while, so maybe we can remove the flags part here? I don't think it's necessary anymore.

Reply:
Possibly.  OpenMP is now generally turned on with RcppArmadillo and packages via the supplied `src/Makevars`. But I am less sure about `cppFunction()` use.